### DISCUSSION TITLE
I'm confused about the name of the 'ref struct'

### DISCUSSION THREAD

**Discussion by paxhujing** [dotnet/runtime#97961] [Q&A] [answered]
'ref struct' looks like it's telling you that the struct is a reference-like type.In effect,it is stuck on the stack.So why not call it 'stack struct'?

**Comment by huoyaoyuan** [comment]
Here the `ref` means the `ref` keyword in `void M(ref int x)`. `ref struct` behaves similar to the parameter x.

**Comment by En3Tho** [comment, marked as answer]
> looks like it's telling you that the struct is a reference-like type

Correct would be "by-reference-like type". It means that compiler should enforce the same rules for such a struct as for example `ref int`.
Living on the stack is an implementation detail/constraint. It could be theoretically possible for a class to have ref field for example. The problems are:
1) making sure that ref field on heap types are only heap referencing or object should be considered dead when functions exits (lifetime issue basically)
2) making sure GC can support arbitrary heap-to-heap refs without a huge perf hit because that ref can point to basically arbitrary field of the object.

So I wouldn't really tie naming to the stack although we might never see heap-to-heap refs and only stack-to-* ones.
